evm/p2p/auth.py

```diff
@@ -0,0 +1,240 @@
+    PUBKEY_LEN,
+    SIGNATURE_LEN,
+    SUPPORTED_RLPX_VERSION,
+)
+from evm.p2p.exceptions import AuthenticationError
+from evm.p2p.utils import (
+    sxor,
+)
+
+
+class HandshakeBase:
+    logger = logging.getLogger("evm.p2p.peer.Handshake")
+    got_eip8_auth = False
+
+    def __init__(self, remote, privkey):
+        self.remote = remote
+        self.privkey = privkey
+        self.ephemeral_privkey = ecies.generate_privkey()
+
+    @property
+    def ephemeral_pubkey(self):
+        return self.ephemeral_privkey.public_key
+
+    @property
+    def pubkey(self):
+        return self.privkey.public_key
+
+    @asyncio.coroutine
+    def connect(self):
+        reader, writer = yield from asyncio.open_connection(
+            host=self.remote.address.ip, port=self.remote.address.tcp_port)
+        return reader, writer
+
+    def derive_secrets(self, initiator_nonce, responder_nonce,
+                       remote_ephemeral_pubkey, auth_init, auth_ack):
+        """Derive base secrets from ephemeral key agreement."""
+        # ecdhe-shared-secret = ecdh.agree(ephemeral-privkey, remote-ephemeral-pubk)
+        ecdhe_shared_secret = ecies.ecdh_agree(
+            self.ephemeral_privkey, remote_ephemeral_pubkey)
+
+        # shared-secret = sha3(ecdhe-shared-secret || sha3(nonce || initiator-nonce))
+        shared_secret = keccak(
+            ecdhe_shared_secret + keccak(responder_nonce + initiator_nonce))
+
+        # aes-secret = sha3(ecdhe-shared-secret || shared-secret)
+        aes_secret = keccak(ecdhe_shared_secret + shared_secret)
+
+        # mac-secret = sha3(ecdhe-shared-secret || aes-secret)
+        mac_secret = keccak(ecdhe_shared_secret + aes_secret)
+
+        # setup sha3 instances for the MACs
+        # egress-mac = sha3.update(mac-secret ^ recipient-nonce || auth-sent-init)
+        mac1 = keccak_updateable(sxor(mac_secret, responder_nonce) + auth_init)
+        # ingress-mac = sha3.update(mac-secret ^ initiator-nonce || auth-recvd-ack)
+        mac2 = keccak_updateable(sxor(mac_secret, initiator_nonce) + auth_ack)
+
+        if self._is_initiator:
+            egress_mac, ingress_mac = mac1, mac2
+        else:
+            egress_mac, ingress_mac = mac2, mac1
+
+        return aes_secret, mac_secret, egress_mac, ingress_mac
+
+
+class HandshakeInitiator(HandshakeBase):
+    _is_initiator = True
+
+    def encrypt_auth_message(self, auth_message):
+        auth_init = ecies.encrypt(auth_message, self.remote.pubkey)
+        return auth_init
+
+    def create_auth_message(self, nonce):
+        ecdh_shared_secret = ecies.ecdh_agree(self.privkey, self.remote.pubkey)
+        secret_xor_nonce = sxor(ecdh_shared_secret, nonce)
+
+        # S(ephemeral-privk, ecdh-shared-secret ^ nonce)
+        S = self.ephemeral_privkey.sign_msg_hash(secret_xor_nonce).to_bytes()
+
+        # S || H(ephemeral-pubk) || pubk || nonce || 0x0
+        auth_message = S + keccak(self.ephemeral_pubkey.to_bytes()) + self.pubkey.to_bytes() + \
```

Comment:
I've found this format a bit more readable than using `\` to split across multiple lines.

```python
auth_message = (
    S +
    keccak(self.ephemeral_pubkey.to_bytes()) +
    self.pubkey.to_bytes() + 
    nonce +
    b'\x00'
)
```

Or:

```python
auth_message = b''.join((
    S,
    keccak(self.ephemeral_pubkey.to_bytes()),
    self.pubkey.to_bytes(),
    nonce,
    b'\x00',
))
```